### DISCUSSION TITLE
Route middlewares as objects, not only strings

### DISCUSSION THREAD

**Discussion by wakebit** [laravel/framework#34056] [Ideas] [answered]
I have suggestion for writing route middlewares more beautiful and understandable.

Example [from docs](https://laravel.com/docs/7.x/authorization#via-middleware) where shown route defining with 'can' middleware:
```php
Route::post('/post', function () {
    // The current user may create posts...
})->middleware('can:create,App\Post');
```

It is hard to read and model is not colorized and not clickable in IDE and code editors.

How it implemented in other places:
[[1]](https://laravel.com/docs/7.x/authorization#via-the-user-model):
```php
$user->can('create', Post::class)
```
[[2]](https://laravel.com/docs/7.x/authorization#via-controller-helpers):
```php
$this->authorize('create', Post::class);
```
[[3]](https://laravel.com/docs/7.x/authorization#via-blade-templates):
```php
@can('create', App\Post::class)
    <!-- The Current User Can Create Posts -->
@endcan
```

Then, if this ability will be implemented for route middlewares, it could look better like this:
```php
Route::post('/post', function () {
    // The current user may create posts...
})->middleware(Can::ability('create', \App\Post::class));
```
or like this:
```php
Route::post('/post', function () {
    // The current user may create posts...
})->middleware(Authorize::ability('create')->parameters(\App\Post::class));
```
or... something else.

Route middlewares dont support objects now, only strings. And [middleware class](https://laravel.com/docs/7.x/middleware#assigning-middleware-to-routes) maybe referenced the only way:
```php 
Route::post('/post', function () {
    // The current user may create posts...
})->middleware(Authorize::class);
```

**Comment by wakebit** [comment, marked as answer]
UPD: Something like this suggestion implemented by @timacdonald 

More detail here: https://timacdonald.me/rethinking-laravels-middleware-argument-api

**Reply by timacdonald** [reply]
Just a heads up that all first party middleware now have static methods on them...

https://github.com/laravel/framework/pull/46362
